File: codewars/python/dirReduc.py
def dirReduc(arr):
    if len(arr) == 0:
        return []
    for i in range(len(arr)-1):
        if arr[i] == "NORTH" and arr[i+1] == "SOUTH" or arr[i+1] == "NORTH" and arr[i] == "SOUTH":
            return dirReduc(arr[:i]+arr[i+2:])
        if arr[i] == "EAST" and arr[i+1] == "WEST" or arr[i+1] == "EAST" and arr[i] == "WEST":
            return dirReduc(arr[:i]+arr[i+2:])
    return arr


if __name__ == '__main__':
    a = ["NORTH", "SOUTH", "SOUTH", "EAST", "WEST", "NORTH", "WEST"]
    print(dirReduc(a))
    u = ["NORTH", "WEST", "SOUTH", "EAST"]
    print(dirReduc(u))

File: codewars/kotlin/dirReduc.kt
fun dirReduc(arr: Array<String>): Array<String> {
    if (arr.isEmpty()) return arr
    for (i in 0 until arr.size - 1) {
        if (arr[i] == "NORTH" && arr[i + 1] == "SOUTH" ||
                arr[i + 1] == "NORTH" && arr[i] == "SOUTH"
        ) {
            return dirReduc(
                arr.slice(0 until i).plus(arr.slice(i + 2 until arr.size)).toTypedArray()
            )
        }
        if (arr[i] == "EAST" && arr[i + 1] == "WEST" ||
                arr[i + 1] == "EAST" && arr[i] == "WEST"
        ) {
            return dirReduc(
                arr.slice(0 until i).plus(arr.slice(i + 2 until arr.size)).toTypedArray()
            )
        }
    }
    return arr
}

fun main() {
    var a = arrayOf("NORTH", "SOUTH", "SOUTH", "EAST", "WEST", "NORTH", "WEST")
    println(dirReduc(a).toList())
}
